qiskit/dagcircuit/dagcircuit.py

```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
             self._decrement_op(node.op)
         return new_node
 
+    def swap_nodes(self, node1, node2):
+        """Swap connected nodes e.g. due to commutation.
+
+        Args:
+            node1 (OpNode): predecessor node
+            node2 (OpNode): successor node
+
+        Raises:
+            DAGCircuitError: if either node is not an OpNode
+        """
+        if not (isinstance(node1, DAGOpNode) and isinstance(node2, DAGOpNode)):
+            raise DAGCircuitError("nodes to swap are not both OpNodes")
+        try:
+            connected_edges = self._multi_graph.get_all_edge_data(node1._node_id, node2._node_id)
+        except rx.NoEdgeBetweenNodes:
+            # since nodes aren't connected there's nothing to do
+            return
+        node1_id = node1._node_id
+        node2_id = node2._node_id
+        edge_find = lambda x: x == edge
+        for edge in connected_edges[::-1]:
+            edge_parent = self._multi_graph.find_predecessors_by_edge(node1_id, edge_find)[0]
+            self._multi_graph.remove_edge(edge_parent._node_id, node1_id)
+            self._multi_graph.add_edge(edge_parent._node_id, node2_id, edge)
+            edge_child = self._multi_graph.find_successors_by_edge(node2_id, lambda x: x == edge)[0]
```

Comment:
I don't know if there's an equivalent for `predecessors`, but this method could be `find_adjacent_node_by_edge` to avoid calculating the whole list.

Reply:
How do you make `find_adjacent_node_by_edge` filter on predecessor or successor?